File: 백준/Silver/9012. 괄호/괄호.py
def check(exp):
    stack = []
    for x in exp:
        if x == '(':
            stack.append(x)
        else:
            if not stack:
                return 0
            ch = stack.pop()
            while ch != '(':
                ch = stack.pop()
    if stack:
        return 0
    else:
        return 1


for _ in range(int(input())):
    a = input()
    if check(a) == 0:
        print('NO')
    else:
        print('YES')

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        for(int i =0;i<n;i++){
            String input = br.readLine();
            if(check(input)){
                System.out.println("YES");
            }
            else{
                System.out.println("NO");
            }
        }
    }
    public static boolean check(String input){
        Stack<Character> stack = new Stack<>();
        for(int i=0;i<input.length();i++){
            char ch = input.charAt(i);
            if(ch == '('){
                stack.push(ch);
            }
            else if(ch == ')'){
                if(stack.isEmpty()){
                    return false;
                }
                char open_ch = stack.pop();
                if(open_ch != '('){
                    return false;
                }
            }

        }
        if(!stack.isEmpty()){
            return false;
        }
        return true;
    }
}